File: Tema 10/ListadoPersonasClean/Domain/UseCases/PersonaUseCase.cs
using Domain.DTOS;
using Domain.Entities;
using Domain.Interfaces;
using Domain.Repositorios;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Domain.UseCases
{
    public class PersonaUseCase : IPersonaUseCase
    {
        private readonly IPersonaRepositorio _personaRepositorio;
        private readonly IDepartamentoRepositorio _departamentoRepositorio;

        public PersonaUseCase(IPersonaRepositorio personaRepositorio, IDepartamentoRepositorio departamentoRepositorio)
        {
            _personaRepositorio = personaRepositorio;
            _departamentoRepositorio = departamentoRepositorio;
        }

        public List<PersonaConNombreDepartamento> getPersonas()
        {
            var personas = _personaRepositorio.getPersonas();
            var departamentos = _departamentoRepositorio.getDepartamentos();

            var listado = new List<PersonaConNombreDepartamento>();

            foreach (var persona in personas)
            {
                var departamento = departamentos.FirstOrDefault(d => d.Id == persona.IdDepartamento);
                string nombreDepartamento = departamento?.Nombre ?? "No tiene departamento";

                listado.Add(new PersonaConNombreDepartamento(persona, nombreDepartamento));
            }

            return listado;
        }

        public PersonaConNombreDepartamento getPersona(int id)
        {
            var persona = _personaRepositorio.getPersona(id);
            if (persona == null) return null;

            var departamento = _departamentoRepositorio.getDepartamento(persona.IdDepartamento ?? 0);
            string nombreDepartamento = departamento?.Nombre ?? "No tiene departamento";

            return new PersonaConNombreDepartamento(persona, nombreDepartamento);
        }

        public PersonaConListadoDepartamento getPersonaEditar(int id)
        {
            var persona = _personaRepositorio.getPersona(id);

            var departamentos = _departamentoRepositorio.getDepartamentos();

            return new PersonaConListadoDepartamento(persona, departamentos);
        }

        public PersonaConListadoDepartamento getPersonaAgregar()
        {
            var persona = new Persona();

            var departamentos = _departamentoRepositorio.getDepartamentos();

            return new PersonaConListadoDepartamento(persona, departamentos);
        }

        public List<Departamento> getDepartamentos()
        {
            return _departamentoRepositorio.getDepartamentos();
        }

        public int agregarPersona(Persona persona)
        {
            return _personaRepositorio.agregarPersona(persona);
        }

        public int actualizarPersona(Persona persona)
        {
            return _personaRepositorio.actualizarPersona(persona);
        }

        public int eliminarPersona(int id)
        {
            return _personaRepositorio.eliminarPersona(id);
        }
    }
}

File: Tema15/ListadoPersonasAngular/src/Domain/UseCases/PersonaUseCase.ts
import { IPersonaRepository } from '../Repositories/IPersonaRepository';
import { IPersonaUseCase } from '../Interfaces/IPersonaUseCase';
import { IDepartamentoRepository } from '../Repositories/IDepartamentoRepository';
import { Persona } from '../Entities/Persona';
import { PersonaDTO } from '../DTOs/PersonaDTO';
import { inject, injectable } from 'inversify';
import { TYPES } from '../../Core/Types';

@injectable()
export class PersonaUseCase implements IPersonaUseCase {
  constructor(
    @inject(TYPES.IPersonaRepository) private personaRepository: IPersonaRepository,
    @inject(TYPES.IDepartamentoRepository) private departamentoRepository: IDepartamentoRepository
  ) {}

  async getPersonas(): Promise<PersonaDTO[]> {
    const personas = await this.personaRepository.getPersonas();
    const departamentos = await this.departamentoRepository.getDepartamentos();
    
    return personas.map(persona => {
      const departamento = departamentos.find(d => d.id === persona.idDepartamento);
      return new PersonaDTO(
        persona.id,
        persona.nombre,
        persona.apellidos,
        persona.fechaNacimiento,
        persona.direccion,
        persona.telefono,
        persona.foto,
        persona.idDepartamento,
        departamento?.nombre || 'Sin departamento'
      );
    });
  }

  async getPersona(id: number): Promise<PersonaDTO> {
    const persona = await this.personaRepository.getPersona(id);
    const departamento = await this.departamentoRepository.getDepartamento(persona.idDepartamento);
    
    return new PersonaDTO(
      persona.id,
      persona.nombre,
      persona.apellidos,
      persona.fechaNacimiento,
      persona.direccion,
      persona.telefono,
      persona.foto,
      persona.idDepartamento,
      departamento?.nombre || 'Sin departamento'
    );
  }

  async agregarPersona(personaDTO: PersonaDTO): Promise<number> {
    const persona = new Persona(
      0,
      personaDTO.nombre,
      personaDTO.apellidos,
      personaDTO.fechaNacimiento,
      personaDTO.direccion,
      personaDTO.telefono,
      personaDTO.foto,
      personaDTO.idDepartamento
    );
    
    return await this.personaRepository.agregarPersona(persona);
  }

  async actualizarPersona(personaDTO: PersonaDTO): Promise<number> {
    const persona = new Persona(
      personaDTO.id,
      personaDTO.nombre,
      personaDTO.apellidos,
      personaDTO.fechaNacimiento,
      personaDTO.direccion,
      personaDTO.telefono,
      personaDTO.foto,
      personaDTO.idDepartamento
    );
    
    return await this.personaRepository.actualizarPersona(persona);
  }

  async eliminarPersona(id: number): Promise<number> {
    return await this.personaRepository.eliminarPersona(id);
  }
}